attach.go

```diff
@@ -0,0 +1,155 @@
+package mail
+
+import (
+	"encoding/base64"
+	"errors"
+	"io/ioutil"
+	"mime"
+	"path/filepath"
+)
+
+// File represents the file that can be added to the email message.
+// You can add attachment from file in path, from base64 string or from []byte.
+// You can define if attachment is inline or not.
+type File struct {
+	// FilePath is the path of the file to attach.
+	FilePath string
+	// Name is the name of file in attachment. Required for Data and B64File. Optional for FilePath.
+	Name string
+	// MimeType of attachment. If empty then is obtained from Name (if not empty) or FilePath. If cannot obtained, application/octet-stream is set.
+	MimeType string
+	// B64File is the base64 string of file to attach.
+	B64File string
+	// Data is the []byte of file to attach.
+	Data []byte
+	// Inline defines if attachment is inline or not.
+	Inline bool
+}
+
+type attachType int
+
+const (
+	attachData attachType = iota
+	attachB64
+	attachFile
+)
+
+// Attach allows you to add an attachment to the email message.
+// The attachment can be inlined
+func (email *Email) Attach(file *File) *Email {
+	if email.Error != nil {
+		return email
+	}
+
+	var name = file.Name
+	var mimeType = file.MimeType
+
+	// if no alternative name was provided, get the filename
+	if len(name) == 0 && len(file.FilePath) > 0 {
+		_, name = filepath.Split(file.FilePath)
+	}
+
+	// get the mimetype
+	if mimeType == "" {
+		mimeType = mime.TypeByExtension(filepath.Ext(name))
+		if mimeType == "" {
+			mimeType = "application/octet-stream"
+		}
+	}
+
+	attachTy, err := getAttachmentType(file)
+	if err != nil {
+		email.Error = errors.New("Mail Error: Failed to add attachment with following error: " + err.Error())
+		return email
+	}
+
+	file.Name = name
+	file.MimeType = mimeType
+
+	switch attachTy {
+	case attachData:
+		email.attachData(file)
+	case attachB64:
+		email.Error = email.attachB64(file)
+	case attachFile:
+		email.Error = email.attachFile(file)
+	}
+
+	return email
+}
+
+func getAttachmentType(file *File) (attachType, error) {
+	// 1- data
+	// 2- base64
+	// 3- file
+
+	// first check if Data
+	if len(file.Data) > 0 {
+		// data requires a name
+		if len(file.Name) == 0 {
+			return 0, errors.New("Mail Error: Attach from bytes requires a name")
```

Comment:
I think you can drop the `Mail Error: ` preamble from this error string as you append it on the upper level also.

Reply:
Thanks for catch this. I will correct it.